File: src/app/schemas/carrito.py
from datetime import datetime
from decimal import Decimal
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


class ProductoCarritoBase(BaseModel):
    id: int
    nombre: str
    precio: Decimal
    activo: bool
    tipo_producto: Optional[str] = None
    id_origen: Optional[int] = None

    model_config = ConfigDict(from_attributes=True)


class CarritoBase(BaseModel):
    id: int
    id_usuario: Optional[int] = None
    id_pedido: Optional[int] = None
    token_identificador: Optional[str] = None
    estado: bool
    fecha_creacion: datetime

    model_config = ConfigDict(from_attributes=True)


class CarritoDetalleBase(BaseModel):
    id: int
    id_carrito: int
    id_producto: int
    cant: int
    precio_unitario: Decimal
    subtotal: Decimal
    producto: Optional[ProductoCarritoBase] = None

    model_config = ConfigDict(from_attributes=True)


class CarritoItemAdd(BaseModel):
    id_producto: int
    cant: int = Field(default=1, ge=1)


class CarritoItemUpdate(BaseModel):
    cant: int = Field(ge=1)


class CarritoCheckoutRequest(BaseModel):
    metodo_pago: str = Field(default="transferencia", min_length=2, max_length=50)
    observaciones: Optional[str] = Field(default=None, max_length=1000)


class CarritoCheckoutResponse(BaseModel):
    id_pedido: int
    id_pago: int
    estado_pedido: str
    estado_pago: str
    referencia_externa: Optional[str] = None
    whatsapp_url: str
    total: Decimal
    mensaje: str


class CarritoResumen(BaseModel):
    carrito: CarritoBase
    items: List[CarritoDetalleBase]
    total_unidades: int
    total_importe: Decimal


class CarritoCreate(BaseModel):
    id_usuario: Optional[int] = None
    id_pedido: Optional[int] = None
    token_identificador: str
    estado: bool = True
    fecha_creacion: Optional[datetime] = None


class CarritoPatch(BaseModel):
    id_usuario: Optional[int] = None
    id_pedido: Optional[int] = None
    token_identificador: Optional[str] = None
    estado: Optional[bool] = None
    fecha_creacion: Optional[datetime] = None

File: client/src/types/carrito.ts
export type CarritoProducto = {
  id: number
  nombre: string
  precio: string | number
  activo: boolean
  tipo_producto?: 'equipo' | 'accesorio' | null
  id_origen?: number | null
  foto_url?: string | null
}

export type CarritoDetalle = {
  id: number
  id_carrito: number
  id_producto: number
  cant: number
  precio_unitario: string | number
  subtotal: string | number
  producto: CarritoProducto | null
  /** Unidades vendibles para este producto (tope para cantidad en carrito). */
  stock_disponible?: number
}

export type Carrito = {
  id: number
  id_usuario: number | null
  id_pedido: number | null
  token_identificador: string | null
  estado: boolean
  fecha_creacion: string
}

export type CarritoResumen = {
  carrito: Carrito
  items: CarritoDetalle[]
  total_unidades: number
  total_importe: string | number
}

export type CarritoCheckoutPayload = {
  metodo_pago: string
  observaciones?: string | null
}

export type CarritoCheckoutResponse = {
  id_pedido: number
  id_pago: number
  estado_pedido: string
  estado_pago: string
  referencia_externa: string | null
  whatsapp_url: string
  total: string | number
  mensaje: string
}

export type VarianteTienda = {
  id_producto: number
  color?: string | null
  precio: number
  precio_usd?: number | null
  foto_url?: string | null
  nombre_corto?: string | null
  stock?: number | null
  disponible?: boolean
  atributos?: Record<string, string>
}

export type VarianteAtributoDisponible = {
  code: string
  label: string
  options: string[]
}

export type ProductoCompra = {
  id: number
  nombre: string
  descripcion: string | null
  foto_url?: string | null
  precio: string | number
  precio_usd?: string | number | null
  id_categoria: number
  activo: boolean
  tipo_producto?: 'equipo' | 'accesorio' | null
  id_origen?: number | null
  tipo_equipo?: string | null
  estado_comercial?: string | null
  /** Catálogo tienda agrupado (backend): variantes por unidad / color */
  tipo_catalogo?: 'grupo_equipo' | string | null
  id_modelo?: number | null
  /** Accesorios: stock numérico en `productos`; equipos: suele venir de variantes */
  stock?: number | null
  variantes_tienda?: VarianteTienda[] | null
}

export type ProductoEquipoDetalle = {
  id_equipo: number
  id_modelo?: number | null
  nombre_modelo?: string | null
  capacidad_gb?: number | null
  color?: string | null
  tipo_equipo?: string | null
  estado_comercial?: string | null
  foto_url?: string | null
}

export type ProductoAccesorioDetalle = {
  id_accesorio: number
  tipo: string
  nombre: string
  color?: string | null
  descripcion?: string | null
  estado: boolean
}

export type ProductoDetalle = ProductoCompra & {
  detalle_equipo?: ProductoEquipoDetalle | null
  detalle_accesorio?: ProductoAccesorioDetalle | null
  variantes_tienda?: VarianteTienda[] | null
  atributos_disponibles?: VarianteAtributoDisponible[] | null
}